src/field_theory/primitive_element.lean

```diff
@@ -0,0 +1,260 @@
+end
+
+/-- Primitive element theorem for finite dimensional extension of a finite field. -/
+theorem primitive_element_fin [fintype F] [hfd : finite_dimensional F E] :
+  ∃ α : E, F⟮α⟯ = ⊤ :=
+begin
+  haveI : fintype E := fintype_of_fintype F E,
+  exact primitive_element_fin_aux F,
+end
+
+end primitive_element_finite
+
+section primitive_element_inf_lemmas
+variables {F : Type*} [field F] {E : Type*} [field E] (ϕ : F →+* E)
+
+/-! ### Primitive element theorem for infinite fields -/
+
+lemma primitive_element_two_inf_exists_c (α β : E) (f g : polynomial F) [F_inf : infinite F] :
+  ∃ c : F, ∀ (α' ∈ (f.map ϕ).roots) (β' ∈ (g.map ϕ).roots), -(α' - α)/(β' - β) ≠ ϕ c :=
+begin
+  let sf := (f.map ϕ).roots,
+  let sg := (g.map ϕ).roots,
+  let s := (sf.bind (λ α', sg.map (λ β', -(α' - α) / (β' - β)))).to_finset,
+  let s' := s.preimage ϕ (λ x hx y hy h, ϕ.injective h),
+  obtain ⟨c, hc⟩ := infinite.exists_not_mem_finset s',
+  simp_rw [finset.mem_preimage, multiset.mem_to_finset, multiset.mem_bind, multiset.mem_map] at hc,
+  push_neg at hc,
+  exact ⟨c, hc⟩,
+end
+
+lemma polynomial.linear_of_splits_single_root {β : F} {h : polynomial F}
+  (h_splits : polynomial.splits ϕ h) (h_roots : (h.map ϕ).roots = {ϕ β}) :
+  h = (polynomial.C (polynomial.leading_coeff h)) * (polynomial.X - polynomial.C β) :=
+begin
+  apply polynomial.map_injective _ ϕ.injective,
+  rw [polynomial.eq_prod_roots_of_splits h_splits,h_roots,multiset.map_singleton,
+      multiset.singleton_eq_singleton,multiset.prod_singleton,polynomial.map_mul,
+      polynomial.map_sub,polynomial.map_X,polynomial.map_C,polynomial.map_C],
+end
+
+lemma polynomial.linear_of_splits_separable_root {β : F} {h : polynomial F} (h_ne_zero : h ≠ 0)
+  (h_sep : h.separable) (h_root : h.eval β = 0) (h_splits : polynomial.splits ϕ h)
+  (h_roots : ∀ x ∈ (h.map ϕ).roots, x = ϕ β) :
+  h = (polynomial.C (polynomial.leading_coeff h)) * (polynomial.X - polynomial.C β) :=
+begin
+  apply polynomial.linear_of_splits_single_root ϕ h_splits,
+  apply finset.mk.inj,
+  { change _ = {ϕ β},
+    rw finset.eq_singleton_iff_unique_mem,
+    split,
+    { apply finset.mem_mk.mpr,
+      rw polynomial.mem_roots (show h.map ϕ ≠ 0, by exact polynomial.map_ne_zero h_ne_zero),
+      rw [polynomial.is_root.def,←polynomial.eval₂_eq_eval_map,polynomial.eval₂_hom,h_root],
+      exact ring_hom.map_zero ϕ },
+    { exact h_roots } },
+  { exact polynomial.nodup_roots (polynomial.separable.map h_sep) },
+end
+
+end primitive_element_inf_lemmas
+
+section primitive_element_theorem
+variables {F : Type*} [field F] {E : Type*} [field E] [algebra F E]
+
+-- This is the heart of the proof of the primitive element theorem. It shows that if `F` is
+-- infinite and `α` and `β` are algebraic over `F` then `F⟮α, β⟯` is generated by a single element.
+lemma primitive_element_two_inf (α β : E) (F_sep : is_separable F E) (F_inf : infinite F) :
+  ∃ γ : E, (F⟮α, β⟯ : set E) ⊆ (F⟮γ⟯ : set E) :=
+begin
```

Comment:
I actually like forwards-style tactic proofs because it's how I think about things when doing mathematics normally. Namely, you build up more and more facts about the context until you can conclude the theorem with what you've built up. But anyway, Thomas said he's working on refactoring this proof so he might follow your suggestion.

Reply:
I did a big refactor which I think is along the lines of your suggestion.